### PR TITLE
Signing request and copy jobs

### PR DESCRIPTION
**Pull request by tbielawa**
Implements signing requests over the UMB and mirroring signatures to internal locations

### CODE REVIEW

File: signing-jobs/build/sign-artifacts/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+node {
+    checkout scm
+    def buildlib = release.buildlib
+    def commonlib = release.commonlib
+
+    // Expose properties for a parameterized build
+    properties(
+        [
+            buildDiscarder(
+                logRotator(
```
Comment (tbielawa) [inline signing-jobs/build/sign-artifacts/Jenkinsfile]: I want these to be infinite retention. Not sure if this is how to do it, but we can fix that in post

File: signing-jobs/build/sign-artifacts/umb_producer.py [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        'amqps://messaging-devops-broker01.web.prod.ext.phx2.redhat.com:5671',
+        'amqps://messaging-devops-broker02.web.prod.ext.phx2.redhat.com:5671',
+    ),
+
+}
+
+TOPIC = 'VirtualTopic.eng.art.artifact.sign'
+
+# TODO: In the future we need to handle 'rhcos' having '4.1'
+# hard-coded into the URL path.
```
Comment (sosiouxme) [inline signing-jobs/build/sign-artifacts/umb_producer.py]: Huh? What exactly needs to handle that?
BTW I should mention. The layout in which we've delivered RHCOS in the past is not the way Clayton wants it. See https://jira.coreos.com/browse/ART-600 for how it's supposed to be at GA. We might want to practice doing it now...

File: signing-jobs/build/mirror-artifacts/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env groovy
+
+node {
+    checkout scm
+    def buildlib = release.buildlib
```
[the remark this answers is not in the document]
Comment (tbielawa) [reply signing-jobs/build/mirror-artifacts/Jenkinsfile]: oh derp. good catch. copy paste fail

File: signing-jobs/build/sign-artifacts/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                    ],
+                    commonlib.mockParam(),
+                ]
+            ],
+            disableConcurrentBuilds()
+        ]
+    )
+
+    commonlib.checkMock()
+    buildlib.cleanWorkdir("${env.WORKSPACE}")
```
[the remark this answers is not in the document]
Comment (tbielawa) [reply signing-jobs/build/sign-artifacts/Jenkinsfile]: :+1:

File: signing-jobs/build/sign-artifacts/umb_producer.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    else:
+        print("Message missing required field: {}".format(validated))
+        exit(1)
+
+    if noop:
+        print("Message we would have sent over the bus:")
+        print(to_send)
+    else:
+        producer = get_bus_producer(env, client_cert, client_key, ca_certs)
+        producer.send_msg({'tester': requestor, 'to': 'test_msg'}, to_send)
```
Comment (tbielawa) [inline signing-jobs/build/sign-artifacts/umb_producer.py]: I just noticed this (also in `json-digest`). I'll ask for suggestions on what to put there instead.

Comment (tbielawa) [reply signing-jobs/build/sign-artifacts/umb_producer.py]: Figured that out.

File: signing-jobs/build/sign-artifacts/umb_producer.py (hunk 2 above)
Comment (tbielawa) [reply signing-jobs/build/sign-artifacts/umb_producer.py]: I do not plan on addressing that in this PR.

File: signing-jobs/build/sign-artifacts/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		// ######################################################################
+		def baseUmbParams = "--requestor ${BUILD_USER} --sig-keyname ${params.KEY_NAME}"
+		baseUmbParams += " --release-name ${params.NAME} --client-cert ${busCertificate}"
+		baseUmbParams += " --client-key ${busKey} --env ${params.ENV}"
+		baseUmbParams += "${noop}"
+
+		// ######################################################################
+		def openshiftJsonSignParams = "${baseUmbParams}"
+		openshiftJsonSignParams += "--product openshift"
+		openshiftJsonSignParams += "--request-id 'openshift-json-digest ${env.BUILD_URL}"
```
[the remark this answers is not in the document]
Comment (sosiouxme) [reply signing-jobs/build/sign-artifacts/Jenkinsfile]: Also I find all this string addition distasteful (makes mistakes like this easy). And the repetition of the var. I would probably write it like so:
```
def openshiftJsonSignParams = buildlib.cleanWhitespace("""
    ${baseUmbParams}
    --product openshift
    --request-id 'openshift-json-digest ${env.BUILD_URL}'
""")
```
￼
alternatively, put the `cleanWhitespace` in the `commonlib.shell` invocation below.

File: signing-jobs/build/sign-artifacts/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		echo "Submitting OpenShift Payload JSON claim signature request"
+		res = commonlib.shell(
+		    returnAll: true,
+		    script: "./umb_producer.py json-digest ${openshiftJsonSignParams}"
+		)
+
+		// ######################################################################
+		def openshiftSha256SignParams = "${baseUmbParams}"
+		openshiftSha256SignParams += "--product openshift"
+		openshiftSha256SignParams += "--request-id 'openshift-message-digest ${env.BUILD_URL}"
```
Comment (sosiouxme) [inline signing-jobs/build/sign-artifacts/Jenkinsfile]: same problem these two lines

File: signing-jobs/build/sign-artifacts/umb_producer.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+    """Get the releasestream metadata for the payload.
+
+    :return: A `dict` of the JSON at the release stream endpoint
+"""
+    res = requests.get(PAYLOAD_META,
+                       verify=ssl.get_default_verify_paths().openssl_cafile)
+    if res.status_code == 200:
+        return res.json()
+    else:
+        raise(Exception(res.reason))
```
Comment (sosiouxme) [inline signing-jobs/build/sign-artifacts/umb_producer.py]: nit:
```suggestion
        raise Exception(res.reason)
```

File: signing-jobs/build/sign-artifacts/umb_producer.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+
+def presend_validation(message):
+    """Verify the message we want send over the bus has all the required
```
Comment (sosiouxme) [inline signing-jobs/build/sign-artifacts/umb_producer.py]:
```suggestion
    """Verify the message we want to send over the bus has all the required fields"""
```

File: signing-jobs/build/sign-artifacts/umb_producer.py [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+call
+    """
+    image_info_raw = subprocess.check_output(
+        ['oc', 'image', 'info', '-o', 'json', pullspec])
+    return json.loads(image_info_raw)
+
+
+def get_bus_producer(env, certificate, private_key, trusted_certificates):
+    """This is just a wrapper around creating a producer. We're going to
+do need this in multiple places though, so we want to ensure we do it
```
Comment (sosiouxme) [inline signing-jobs/build/sign-artifacts/umb_producer.py]:
```suggestion
need this in multiple places though, so we want to ensure we do it
```
indent?

File: signing-jobs/build/sign-artifacts/Jenkinsfile (hunk 1 above)
Comment (tbielawa) [reply signing-jobs/build/sign-artifacts/Jenkinsfile]: nice.

File: signing-jobs/build/sign-artifacts/Jenkinsfile [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+		// ######################################################################
+		def openshiftJsonSignParams = buildlib.cleanWhitespace("""
+		    ${baseUmbParams}
+		    --product openshift
+		    --request-id 'openshift-json-digest ${env.BUILD_URL}'
+                """)
+
+		echo "Submitting OpenShift Payload JSON claim signature request"
+		res = commonlib.shell(
+		    returnAll: true,
```
Comment (sosiouxme) [inline signing-jobs/build/sign-artifacts/Jenkinsfile]: This will always succeed. Shouldn't we check the `returnStatus`?

Comment (tbielawa) [reply signing-jobs/build/sign-artifacts/Jenkinsfile]: That's a good point!

**Pull request merged**